File: src/services/indicators.ts
import type { OHLCVCandle } from './candleService';

export interface TimeValue {
  time: number;
  value: number;
}

export interface MACDPoint {
  time: number;
  macd: number;
  signal: number;
  histogram: number;
}

export interface BollingerPoint {
  time: number;
  upper: number;
  middle: number;
  lower: number;
}

// --- EMA ---
export function calcEMA(candles: OHLCVCandle[], period: number): TimeValue[] {
  if (candles.length < period) return [];
  const k = 2 / (period + 1);
  const result: TimeValue[] = [];
  let ema = 0;

  // Seed with SMA of first `period` closes
  for (let i = 0; i < period; i++) ema += candles[i].close;
  ema /= period;
  result.push({ time: candles[period - 1].time, value: ema });

  for (let i = period; i < candles.length; i++) {
    ema = candles[i].close * k + ema * (1 - k);
    result.push({ time: candles[i].time, value: ema });
  }
  return result;
}

// --- RSI ---
export function calcRSI(candles: OHLCVCandle[], period = 14): TimeValue[] {
  if (candles.length <= period) return [];
  const result: TimeValue[] = [];
  let avgGain = 0;
  let avgLoss = 0;

  for (let i = 1; i <= period; i++) {
    const delta = candles[i].close - candles[i - 1].close;
    if (delta > 0) avgGain += delta;
    else avgLoss -= delta;
  }
  avgGain /= period;
  avgLoss /= period;

  const rsi0 = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  result.push({ time: candles[period].time, value: rsi0 });

  for (let i = period + 1; i < candles.length; i++) {
    const delta = candles[i].close - candles[i - 1].close;
    const gain = delta > 0 ? delta : 0;
    const loss = delta < 0 ? -delta : 0;
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
    const rsi = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
    result.push({ time: candles[i].time, value: rsi });
  }
  return result;
}

// --- MACD ---
export function calcMACD(
  candles: OHLCVCandle[],
  fast = 12,
  slow = 26,
  signal = 9
): MACDPoint[] {
  const fastEMA = calcEMA(candles, fast);
  const slowEMA = calcEMA(candles, slow);

  // Align by time
  const slowMap = new Map(slowEMA.map(p => [p.time, p.value]));
  const macdLine: TimeValue[] = fastEMA
    .filter(p => slowMap.has(p.time))
    .map(p => ({ time: p.time, value: p.value - slowMap.get(p.time)! }));

  if (macdLine.length < signal) return [];

  // Signal = EMA(signal) of macdLine
  const k = 2 / (signal + 1);
  let sig = 0;
  for (let i = 0; i < signal; i++) sig += macdLine[i].value;
  sig /= signal;

  const result: MACDPoint[] = [];
  result.push({
    time: macdLine[signal - 1].time,
    macd: macdLine[signal - 1].value,
    signal: sig,
    histogram: macdLine[signal - 1].value - sig,
  });

  for (let i = signal; i < macdLine.length; i++) {
    sig = macdLine[i].value * k + sig * (1 - k);
    result.push({
      time: macdLine[i].time,
      macd: macdLine[i].value,
      signal: sig,
      histogram: macdLine[i].value - sig,
    });
  }
  return result;
}

// --- Bollinger Bands ---
export function calcBollinger(
  candles: OHLCVCandle[],
  period = 20,
  stdDev = 2
): BollingerPoint[] {
  if (candles.length < period) return [];
  const result: BollingerPoint[] = [];

  for (let i = period - 1; i < candles.length; i++) {
    let sum = 0;
    for (let j = i - period + 1; j <= i; j++) sum += candles[j].close;
    const mean = sum / period;

    let variance = 0;
    for (let j = i - period + 1; j <= i; j++) {
      const diff = candles[j].close - mean;
      variance += diff * diff;
    }
    const sd = Math.sqrt(variance / period);

    result.push({
      time: candles[i].time,
      upper: mean + stdDev * sd,
      middle: mean,
      lower: mean - stdDev * sd,
    });
  }
  return result;
}

export interface IndicatorData {
  ema21: TimeValue[];
  ema50: TimeValue[];
  ema200: TimeValue[];
  rsi: TimeValue[];
  macd: MACDPoint[];
  bollinger: BollingerPoint[];
}

export function computeAllIndicators(candles: OHLCVCandle[]): IndicatorData {
  return {
    ema21:     calcEMA(candles, 21),
    ema50:     calcEMA(candles, 50),
    ema200:    calcEMA(candles, 200),
    rsi:       calcRSI(candles, 14),
    macd:      calcMACD(candles, 12, 26, 9),
    bollinger: calcBollinger(candles, 20, 2),
  };
}

File: trading_system/indicators.py
"""
IndicatorLibrary: Computes technical indicators on OHLCV DataFrames.

Expected DataFrame columns (case-insensitive): open, high, low, close, volume
Returns a new DataFrame with the original columns plus indicator columns appended.
"""
from __future__ import annotations

import numpy as np
import pandas as pd

from .models import IndicatorConfig, IndicatorType


def _normalize_columns(df: pd.DataFrame) -> pd.DataFrame:
    df = df.copy()
    df.columns = [c.lower() for c in df.columns]
    required = {"open", "high", "low", "close"}
    missing = required - set(df.columns)
    if missing:
        raise ValueError(f"DataFrame missing columns: {missing}")
    if "volume" not in df.columns:
        df["volume"] = 0.0
    return df


def _rsi(close: pd.Series, period: int = 14) -> pd.Series:
    delta = close.diff()
    gain = delta.clip(lower=0)
    loss = -delta.clip(upper=0)

    avg_gain = gain.ewm(com=period - 1, min_periods=period).mean()
    avg_loss = loss.ewm(com=period - 1, min_periods=period).mean()

    rs = avg_gain / avg_loss.replace(0, np.nan)
    return 100 - (100 / (1 + rs))


def _ema(close: pd.Series, period: int) -> pd.Series:
    return close.ewm(span=period, adjust=False).mean()


def _sma(close: pd.Series, period: int) -> pd.Series:
    return close.rolling(window=period).mean()


def _macd(
    close: pd.Series,
    fast: int = 12,
    slow: int = 26,
    signal: int = 9,
) -> tuple[pd.Series, pd.Series, pd.Series]:
    fast_ema = _ema(close, fast)
    slow_ema = _ema(close, slow)
    macd_line = fast_ema - slow_ema
    signal_line = _ema(macd_line, signal)
    histogram = macd_line - signal_line
    return macd_line, signal_line, histogram


def _bollinger_bands(
    close: pd.Series,
    period: int = 20,
    std: float = 2.0,
) -> tuple[pd.Series, pd.Series, pd.Series]:
    middle = _sma(close, period)
    rolling_std = close.rolling(window=period).std()
    upper = middle + std * rolling_std
    lower = middle - std * rolling_std
    return upper, middle, lower


def _atr(high: pd.Series, low: pd.Series, close: pd.Series, period: int = 14) -> pd.Series:
    tr = pd.concat(
        [
            high - low,
            (high - close.shift()).abs(),
            (low - close.shift()).abs(),
        ],
        axis=1,
    ).max(axis=1)
    return tr.ewm(com=period - 1, min_periods=period).mean()


def _stochastic(
    high: pd.Series,
    low: pd.Series,
    close: pd.Series,
    k_period: int = 14,
    d_period: int = 3,
) -> tuple[pd.Series, pd.Series]:
    lowest_low = low.rolling(window=k_period).min()
    highest_high = high.rolling(window=k_period).max()
    k = 100 * (close - lowest_low) / (highest_high - lowest_low).replace(0, np.nan)
    d = k.rolling(window=d_period).mean()
    return k, d


def _cci(high: pd.Series, low: pd.Series, close: pd.Series, period: int = 20) -> pd.Series:
    typical_price = (high + low + close) / 3
    mean_tp = typical_price.rolling(window=period).mean()
    mean_deviation = typical_price.rolling(window=period).apply(
        lambda x: np.mean(np.abs(x - x.mean())), raw=True
    )
    return (typical_price - mean_tp) / (0.015 * mean_deviation.replace(0, np.nan))


def compute_indicators(df: pd.DataFrame, config: IndicatorConfig) -> pd.DataFrame:
    """
    Compute technical indicators and append columns to the DataFrame.

    Parameters
    ----------
    df : pd.DataFrame
        OHLCV DataFrame with at minimum open, high, low, close columns.
    config : IndicatorConfig
        Indicator type and parameters to compute.

    Returns
    -------
    pd.DataFrame
        Original DataFrame with indicator columns appended.

    Raises
    ------
    ValueError
        If required columns are missing or indicator type is unsupported.
    """
    df = _normalize_columns(df)
    p = config.params
    itype = config.indicator_type

    if itype == IndicatorType.RSI:
        df["rsi"] = _rsi(df["close"], period=p.get("period", 14))

    elif itype == IndicatorType.MACD:
        macd_line, signal_line, histogram = _macd(
            df["close"],
            fast=p.get("fast", 12),
            slow=p.get("slow", 26),
            signal=p.get("signal", 9),
        )
        df["macd"] = macd_line
        df["macd_signal"] = signal_line
        df["macd_hist"] = histogram

    elif itype == IndicatorType.EMA_CROSS:
        df["ema_fast"] = _ema(df["close"], period=p.get("fast", 9))
        df["ema_slow"] = _ema(df["close"], period=p.get("slow", 21))

    elif itype == IndicatorType.SMA_CROSS:
        df["sma_fast"] = _sma(df["close"], period=p.get("fast", 50))
        df["sma_slow"] = _sma(df["close"], period=p.get("slow", 200))

    elif itype == IndicatorType.BOLLINGER:
        upper, middle, lower = _bollinger_bands(
            df["close"],
            period=p.get("period", 20),
            std=p.get("std", 2.0),
        )
        df["bb_upper"] = upper
        df["bb_middle"] = middle
        df["bb_lower"] = lower

    elif itype == IndicatorType.ATR:
        df["atr"] = _atr(
            df["high"], df["low"], df["close"], period=p.get("period", 14)
        )

    elif itype == IndicatorType.STOCHASTIC:
        k, d = _stochastic(
            df["high"],
            df["low"],
            df["close"],
            k_period=p.get("k_period", 14),
            d_period=p.get("d_period", 3),
        )
        df["stoch_k"] = k
        df["stoch_d"] = d

    elif itype == IndicatorType.CCI:
        df["cci"] = _cci(
            df["high"], df["low"], df["close"], period=p.get("period", 20)
        )

    else:
        raise ValueError(f"Unsupported indicator type: {itype}")

    return df
